Fix: Handle base[href] tags

chrisguttandin opened the pull request:
<!--

Read our pull request guide:
https://webhint.io/docs/contributor-guide/getting-started/pull-requests/

For the following items put an "x" between the square brackets
(i.e. [x]) if you completed the associated item.

-->

## Pull request checklist

Make sure you:

- [x] Signed the [Contributor License Agreement](https://cla.js.foundation/webhintio/hint)
- [x] Followed the [commit message guidelines](https://webhint.io/docs/contributor-guide/getting-started/pull-requests/#commit-messages)

For non-trivial changes, please make sure you also:

- [ ] Added/Updated related documentation.
- [x] Added/Updated related tests.

## Short description of the change(s)

I discovered a bug in the "no-broken-links" hint, which causes it to ignore a base tag with an href value when resolving relative URLs.

I put together a fix which works when testing it with a site hosted on GitHub pages. However I was not able to add any tests as I couldn't get the tests to work in the first place. I followed the guide on the website. (https://webhint.io/docs/contributor-guide/getting-started/development-environment/)

Are the tests currently broken or am I doing something wrong?

alrra commented:
> Are the tests currently broken or am I doing something wrong?

Did you run `yarn test` from the root of the project? And if so, what exactly was the error? Thanks!

chrisguttandin commented:
Sorry, I forgot to run `yarn` first inside of the directory of the package. Running `yarn test` works in the root directory and in the directory of the package.

I will try to add some tests for my changes.

alrra commented:
> Sorry, I forgot to run yarn first inside of the directory of the package

Sorry for the confusion, we need to improve those docs.

> I will try to add some tests for my changes.

@chrisguttandin Thanks!

chrisguttandin commented:
I refactored my changes a bit and also added two tests.

The rule does now cache all discovered elements with their urls. Once `traverse::end` is fired it will look for the first base tag and will use this (if present) to check all the cached urls. I tried to write that code in a similar way as the respective part of the `hint-meta-charset-utf-8` rule is written.

This behavior relies on the `element::*` events to always happen before the `traverse::end` event and also requires the `traverse::end` event to always fire. Is that guaranteed or should the code be written in way which accounts for a different order of events.

molant commented:
Thanks for your PR @chrisguttandin ! I'll review it today.

> This behavior relies on the `element::*` events to always happen before the `traverse::end` event and also requires the traverse::end event to always fire. Is that guaranteed or should the code be written in way which accounts for a different order of events.

If there's an HTML then that's the right order of events so this should be fine.

molant commented on packages/hint-no-broken-links/src/hint.ts:
```diff
@@ -169,15 +146,56 @@ export default class NoBrokenLinksHint implements IHint {
+
+            const reports: Array<Promise<void>> = collectedElementsWithUrls.reduce<Promise<void>[]>((accumulatedReports, [element, urls]) => {
+                return [...accumulatedReports, ...urls.map(async (url) => {
+                    const isRelativeUrl = (!url.startsWith('/') && URL.parse(url).hostname === null);
```
I believe this is part of the [Legacy URL API](https://nodejs.org/dist/latest-v8.x/docs/api/url.html#url_legacy_url_api). It seems we didn't catch this on the original PR ☹

chrisguttandin replied:
I changed the code now to use the new WHATWG URL API instead. I initially thought I should better keep it in the way it was written before.

molant replied:
And you were right on doing that. On the original PR for this rule we didn't realize we were using the old API. Sorry about that and thanks for updating it!

molant commented on packages/hint-no-broken-links/src/hint.ts:
```diff
@@ -169,15 +146,56 @@ export default class NoBrokenLinksHint implements IHint {
+            const hrefAttribute = (baseTags.length === 0) ? null : baseTags[0].getAttribute('href');
+
+            const reports: Array<Promise<void>> = collectedElementsWithUrls.reduce<Promise<void>[]>((accumulatedReports, [element, urls]) => {
+                return [...accumulatedReports, ...urls.map(async (url) => {
```
Is there any particular reason of using spreading instead of `concat`? Stylistic is a good answer btw.

sarvaje replied:
IMO `urls.map` should be move to its own variable/method and then `return [...accumulatedReports, ...reports]` or even `return [...accumulatedReport, ...getReports(urls)]`.
I think it would be more readable.

chrisguttandin replied:
That was just a stylistic decision. Let me know if you want me to change it to use `concat()` instead.

molant replied:
Maybe just do what @sarvaje proposes of extracting `urls.map` to its own variable to make it a bit clearer 😊

molant commented on packages/hint-no-broken-links/src/hint.ts:
```diff
@@ -169,15 +146,56 @@ export default class NoBrokenLinksHint implements IHint {
+                        }
+                    } else {
+                        // An element which was not present in the fetch end results
+                        return await requester
```
What do you think about removing the `async/await` from this function. I know it's in the original code but I think it can be safely removed.

chrisguttandin replied:
Yes, you're right. There seems to be no reason for awaiting that promise.

molant commented on packages/hint-no-broken-links/src/hint.ts:
```diff
@@ -42,6 +42,9 @@ export default class NoBrokenLinksHint implements IHint {
         const requester = new Requester(options);
         const brokenStatusCodes = [404, 410, 500, 503];
 
+        /** Stores the elements with their urls which have been collected while traversing the page. */
```
It should be `URLs` (you can see @alrra complaining about this in virtually any PR we do 😅). Can you please update it in here and any other comment instance in this file? I think our policy for `URL` in variables isn't clear. @alrra, do you want to update the code? Things like `fetchedUrls`, `urls` (eslint will complain about upper case and constructors)?

alrra replied:
Maybe start using `somethingURL` / `urlSomething`?

chrisguttandin replied:
I'm a bit confused. Should I change the spelling of "URL" only in the comments or should I rename all variables?

alrra replied:
> I change the spelling of "URL" only in the comments or should I rename all variables?

Both would be great! Thanks!

chrisguttandin replied:
Ok, no problem. And what should be the new name of the variable called `url` and `urls`? Should they be renamed to something like `uRL`?

alrra replied:
> variable called `url` and `urls`

@chrisguttandin They should remain the same. Thanks again!

molant commented on packages/hint-no-broken-links/tests/tests.ts:
```diff
@@ -71,6 +75,20 @@ const tests: Array<HintTest> = [
+        }
+    },
+    {
+        name: `This test should pass as it has a link a valid link (when resolved with the base tag)`,
```
I think the name for this test is not correct. Can you please update it?

molant commented on packages/hint-no-broken-links/tests/tests.ts:
```diff
@@ -71,6 +75,20 @@ const tests: Array<HintTest> = [
+            '/': {content: generateHTMLPage('<base href="nested/">', bodyWithValidRelativeLink)},
+            '/nested/about': {content: 'My about page content'}
+        }
+    },
```
Maybe add a test that fails? Something like:

```js
{
  name: `This test should fail because resolved URL with base doesn't exist `,
  reports: [{message: `Broken link found (404 response).` }],
  serverConfig: {
    '/': {content: generateHTMLPage('<base href="nested/">', bodyWithValidRelativeLink)},
    '/nested/about': {status: 404}
  }
}
```

I haven't tested the code above so there might be some modifications to do. There's more information in the [contributor guide](https://webhint.io/docs/contributor-guide/how-to/test-rules/). Also feel free to change the name of the test 😊

alrra commented:
@chrisguttandin Fix is now available in [`@hint/hint-no-broken-links v1.0.3`](https://www.npmjs.com/package/@hint/hint-no-broken-links). 🎉

chrisguttandin commented:
@alrra Thanks a lot.
